### ISSUE TITLE
server: support leveled logging

### ISSUE THREAD

**Issue by ericchiang** [coreos/dex#677]

**Comment by ericchiang** [comment]
This may be a good time to switch to https://github.com/Sirupsen/logrus. The biggest question is if we use a global logger or a struct specific one (i'm in favor of the latter).

```go
type Server struct {
    // Setup during call to NewServer. Defaults to error debug level.
    logger logrus.Logger
}

func (s *Server) debugf(format string, a ...interface{}) { s.logger.Debugf(format, a...) }
func (s *Server) infof(format string, a ...interface{})  { s.logger.Infof(format, a...) }
func (s *Server) errorf(format string, a ...interface{}) { s.logger.Errorf(format, a...) }
```

Probably the only levels we need to support.

Steps.
1. Replace all of our current log calls with `errorf` calls.
2. Centralize all error logging in the server package (remove other logging in the storage package).
3. Add lots of info calls.
4. Add debug calls for.
  * Every storage access
  * Every http request

**Comment by ericchiang** [comment]
cc @rithujohn191

**Comment by ericchiang** [comment]
On second thought, storages need to be able to log things as well. We should probably pass a logger as part of storage creation.

**Comment by ericchiang** [comment]
Per conversation with @chancez we should support the JSON formatter too for log aggregation and filtering.

To expand on the storage needing logging we want basically everything to embed a logger. Storages, connectors, the server, etc. _Please, no global logger._.

Probably want something like the following somewhere in `cmd/dex`

```go
var level logrus.Level
switch config.Debug.Level {
case "debug":
    level = logrus.DebugLevel
case "", "info": // default to info
    level = logrus.InfoLevel
case "error":
    level = logrus.ErrorLevel
default:
    // handle error
}

var formatter logrus.Formatter
switch config.Debug.Format {
case "", "text": // default to text
    formatter = &logrus.TextFormatter{DisableColors: true}
case "json":
    formatter = &logrus.JSONFormatter{}
default:
    // handle error
}

logger := &Logger{
  Out:       os.Stderr,
  Formatter: formatter,
  Level:     level,
}
```

Then as we pass the logger to various components we want to add a new field

```
serverLogger := logger.WithField("component", "server")
```

Or if the server spins of a logger to the connectors.

```
connLogger := serverLogger.WithField("component", conn.ID)
```

Though maybe we can find something shorter than `"component"`?

**Comment by chancez** [comment]
That all looks great to me. You can use this https://godoc.org/github.com/Sirupsen/logrus#ParseLevel to get the log level from a string into a the level object though.

**Comment by chancez** [comment]
I also recommend you use this https://godoc.org/github.com/Sirupsen/logrus#FieldLogger as the object you use everywhere, rather than `logrus.Logger`, since that will allow you to use both `logrus.Logger` and `logrus.Entry` as your logger.

**Comment by ericchiang** [comment]
> You can use this https://godoc.org/github.com/Sirupsen/logrus#ParseLevel to get the log level from a string into a the level object though.

Thanks, though we don't want to support certain logging levels like panic, fatal, and warn. A switch statement is easy enough.

> I also recommend you use this https://godoc.org/github.com/Sirupsen/logrus#FieldLogger as the object you use everywhere, rather than logrus.Logger, since that will allow you to use both logrus.Logger and logrus.Entry as your logger.

Yeah sounds good. Will update my examples.

**Comment by ericchiang** [comment]
closed via #723
